File: packages/contracts/src/relay_contracts/engine.py
"""CEL engine-selection factory -- the SINGLE RELAY_CEL_ENGINE read site.

``make_cel_evaluator()`` is the one place in the whole codebase that reads the
``RELAY_CEL_ENGINE`` environment variable. As of milestone M6 (WS-I) the
single wasm CEL engine is the ONLY Python CEL backend:

  - ``wasm`` (the default, and the value when ``RELAY_CEL_ENGINE`` is unset
    or blank): :class:`relay_contracts.wasm_backed_evaluator.WasmCelEvaluator`.
  - ANY other value -- including the legacy engine token that served as the
    M5 one-release rollback escape hatch -- FAILS CLOSED with a clear
    :class:`ValueError` naming the bad value AND the (wasm-only) allowed set.
    The rollback hatch is closed: the legacy Python CEL engine and its
    dependency were removed at M6, so an explicit legacy selection must be a
    loud structured failure, never a silent fallback to a default.

Why selection lives ONLY here (a load-bearing invariant):

  - The gate engine (``packages/gate``) constructs its evaluator through this
    factory (WS-D / VAL-CWC-P2TSGATE-010) and NEVER reads ``RELAY_CEL_ENGINE``
    itself. A gate-src env read would trip the VAL-W8-005 gate-determinism
    grep (the gate decision must not depend on ambient process environment).
  - Keeping the read in exactly one file means the engine surface (wasm-only
    as of M6) is governed at a single auditable point.

History: M1-M4 defaulted to the legacy engine with wasm behind the flag; M5
(WS-H) flipped the default to wasm with the legacy engine selectable as the
bake-window rollback; M6 (WS-I) removed the legacy engine entirely. The env
var itself is retained as the single selection point so a future engine
addition (if ever) lands here and only here.

ASCII-only per CLAUDE.md "ASCII-Safe Source".
"""

from __future__ import annotations

import os
from collections.abc import Iterable
from typing import Any, Protocol, runtime_checkable

from .udf import PureUdf
from .wasm_backed_evaluator import WasmCelEvaluator

# The environment variable that selects the CEL engine. Read in THIS module
# only (asserted by test_engine_factory.test_relay_cel_engine_read_only_in_
# engine_module and the VAL-W8-005 / VAL-CWC-P4DUALRUN-008 determinism greps).
_ENGINE_ENV_VAR: str = "RELAY_CEL_ENGINE"

# Canonical engine tokens. ``wasm`` is the only engine as of M6 (WS-I removed
# the legacy backend). Matching is exact (case-sensitive) after
# surrounding-whitespace trim -- no locale-dependent case-folding, so
# selection is deterministic.
_ENGINE_WASM: str = "wasm"

# The default engine when RELAY_CEL_ENGINE is unset or blank. Flipped to wasm
# at M5 (WS-H); the only engine since M6 (WS-I).
_DEFAULT_ENGINE: str = _ENGINE_WASM

_ALLOWED_ENGINES: tuple[str, ...] = (_ENGINE_WASM,)


@runtime_checkable
class CelEvaluatorProtocol(Protocol):
    """The host-side CEL evaluator facade.

    :class:`WasmCelEvaluator` satisfies this protocol structurally, so the
    factory's declared return type is the protocol rather than the concrete
    class. Downstream consumers (the gate engine in WS-D, ``pipeline.py``)
    type the evaluator they hold as ``CelEvaluatorProtocol`` and stay
    engine-agnostic.

    ``probe_compile`` and ``evaluate_with_trace`` are part of the facade:
    ``pipeline.publish_contract`` uses the former for publish-time
    malformed-syntax rejection through the engine's authoritative compiler,
    and ``pipeline.evaluate_assertion`` uses the latter to reconstruct
    ``udf_outputs_jcs`` / ``udfs_invoked`` from the engine's ``udf_trace``
    forensic field.
    """

    timeout_ms: int

    def compile(self, expression: str) -> Any: ...

    def probe_compile(self, expression: str) -> None: ...

    def evaluate(self, expression: str, bindings: Any = ...) -> Any: ...

    def evaluate_with_trace(
        self, expression: str, bindings: Any = ...
    ) -> tuple[Any, dict[str, list[Any]]]: ...


def _select_engine_name() -> str:
    """Resolve the engine name from ``RELAY_CEL_ENGINE`` (the ONLY env read).

    An absent or blank value resolves to the default (wasm). A non-blank
    value is trimmed of surrounding whitespace (a common shell-export
    accident) and matched case-sensitively against the allowed tokens. An
    unknown value -- including the removed legacy engine token -- raises a
    clear :class:`ValueError` naming the allowed set (fail closed, never a
    silent fallback).
    """
    raw = os.environ.get(_ENGINE_ENV_VAR)
    if raw is None:
        return _DEFAULT_ENGINE
    value = raw.strip()
    if value == "":
        # A set-but-blank env var (e.g. ``RELAY_CEL_ENGINE=``) is the standard
        # "no selection" signal; fall back to the default.
        return _DEFAULT_ENGINE
    if value not in _ALLOWED_ENGINES:
        allowed = ", ".join(repr(name) for name in _ALLOWED_ENGINES)
        raise ValueError(
            f"{_ENGINE_ENV_VAR}={value!r} is not a recognized CEL engine; "
            f"allowed values are {allowed} (unset or blank -> "
            f"{_DEFAULT_ENGINE!r}). Engine names are case-sensitive. The "
            f"legacy engine was removed at M6; the wasm engine is the only "
            f"Python CEL backend."
        )
    return value


def make_cel_evaluator(
    *,
    timeout_ms: int | None = None,
    udfs: Iterable[PureUdf] = (),
) -> CelEvaluatorProtocol:
    """Construct the CEL evaluator for the selected engine.

    Reads ``RELAY_CEL_ENGINE`` (the SINGLE read site). ``wasm`` (the default;
    unset / blank) constructs :class:`WasmCelEvaluator`; any other value
    fails closed in :func:`_select_engine_name`. ``timeout_ms`` and ``udfs``
    are forwarded to the evaluator's constructor with IDENTICAL semantics, so
    the factory is a transparent substitute for constructing the class
    directly: the same bound checks (positive int, <= ``MAX_TIMEOUT_MS``) and
    the same UDF handling (the wasm accepts the 3 native ``relay.*`` UDFs and
    rejects any other fail-closed) apply.

    Args:
        timeout_ms: per-evaluation wall-clock budget in milliseconds. ``None``
            (the default) defers to the evaluator's own default
            (``DEFAULT_TIMEOUT_MS``), so callers that do not care get identical
            behavior to constructing the evaluator with no ``timeout_ms``.
        udfs: pure UDFs to register. Forwarded verbatim. The wasm engine
            rejects any non-allowlist UDF at construction
            (``RelayCelUnsupportedUdfError`` / RELAY-CEL-004).

    Returns:
        A :class:`CelEvaluatorProtocol` (a :class:`WasmCelEvaluator`).

    Raises:
        ValueError: ``RELAY_CEL_ENGINE`` holds an unrecognized value (the
            removed legacy engine token included), OR the forwarded
            ``timeout_ms`` is out of bounds (re-raised from the evaluator
            constructor).
        RelayCelUnsupportedUdfError: a non-allowlist UDF was forwarded to the
            wasm engine.
    """
    engine = _select_engine_name()
    # Only pass timeout_ms when the caller supplied one, so the evaluator's own
    # DEFAULT_TIMEOUT_MS governs the unspecified case (identical to direct
    # construction with no timeout_ms argument).
    kwargs: dict[str, Any] = {"udfs": udfs}
    if timeout_ms is not None:
        kwargs["timeout_ms"] = timeout_ms

    if engine == _ENGINE_WASM:
        return WasmCelEvaluator(**kwargs)
    # Defensive: _select_engine_name only returns an allowed token or raises;
    # this is unreachable. Kept as a fail-closed guard rather than a silent
    # default so a future allowed-set edit that forgets a branch is caught.
    raise ValueError(  # pragma: no cover -- unreachable given _select_engine_name
        f"{_ENGINE_ENV_VAR} resolved to an unhandled engine {engine!r}; "
        f"allowed values are {_ALLOWED_ENGINES}."
    )


__all__ = ["CelEvaluatorProtocol", "make_cel_evaluator"]

File: packages/contracts-typescript/src/engine.ts
// CEL engine-selection factory -- the canonical TS evaluator entry point.
//
// `makeCelEvaluator()` is the TypeScript mirror of the Python factory
// packages/contracts/src/relay_contracts/engine.py (make_cel_evaluator). After
// M6 WS-I the wasm engine is the ONLY TS CEL backend: the legacy engine and
// its rollback escape hatch are removed. The factory therefore has a single
// constructable engine behind the host facade (`compile`, `evaluate`,
// `timeoutMs`):
//
//   - "wasm" (the DEFAULT, and the selection when `engine` is unset or blank):
//     :class WasmCelBackend (wasm-evaluator.ts), the single wasm CEL engine
//     behind the host facade. This is the engine EVERY consumer runs on. The
//     default backend resolves the PACKAGED wasm artifact + `.mjs` loader
//     (WS-G package data) with the explicit precedence explicit > CEL_WASM >
//     packaged data (wasm-artifact.ts resolveWasmPathForLoader /
//     wasm-evaluator.ts defaultLoaderPath), so a fresh install loads the
//     engine with no configuration.
//
// Any OTHER engine name -- including the now-removed legacy spellings -- is
// rejected fail-closed with a clear error naming the bad value AND the
// (wasm-only) allowed set; never a silent fallback to a default. The M5
// rollback hatch (explicit legacy selection) is CLOSED at M6: a deployment can
// no longer select the legacy engine because it does not exist.
//
// DETERMINISM BOUNDARY (load-bearing; differs from Python BY DESIGN): the TS
// selection is CONFIG/PARAM-based, NOT environment-based. The engine-selector
// environment variable (RELAY_CEL_ENGINE) is read ONLY in the Python
// packages/contracts factory (boundaries.md); production TS src/ MUST NOT
// name it -- the AST presence scan in test/default-engine.test.ts FAILS the
// build if any src/ file carries that token as an identifier or string
// literal. A TS consumer that wants env-driven selection performs the env
// read itself (outside this package) and passes the resolved token as
// `engine`. (This comment may name the selector: comments are invisible to
// the AST scan, exactly like the wasm-artifact.ts explanatory comments.)
//
// ASCII-only per CLAUDE.md "ASCII-Safe Source".

import type { PureUdf } from "./udf.js";
import { WasmCelBackend } from "./wasm-evaluator.js";

// Canonical engine token. After M6 WS-I "wasm" is the default AND the only
// constructable engine. Matching is exact (case-sensitive) after surrounding-
// whitespace trim -- no locale-dependent case-folding, so selection is
// deterministic. Mirrors engine.py _ENGINE_WASM.
const ENGINE_WASM = "wasm";

// The default engine when `engine` is unset or blank. After M6 WS-I there is
// no alternative to select. Mirrors engine.py _DEFAULT_ENGINE.
const DEFAULT_ENGINE = ENGINE_WASM;

const ALLOWED_ENGINE_TOKENS: readonly string[] = [ENGINE_WASM];

/**
 * The canonical engine token a TypeScript caller may select. After M6 WS-I
 * `"wasm"` is the sole accepted value.
 */
export type CelEngineName = "wasm";

/**
 * The shared host-side CEL evaluator facade the factory returns: the
 * wasm-backed engine. The TS analogue of the Python CelEvaluatorProtocol
 * (engine.py).
 */
export type CelEvaluator = WasmCelBackend;

export interface MakeCelEvaluatorOptions {
  /**
   * Engine selection token. Unset / blank selects the default (wasm). Any
   * other value -- including the now-removed legacy spellings -- is rejected
   * fail-closed. Typed as `string` (not the literal type) deliberately: this
   * mirrors the Python factory's env-string contract -- arbitrary runtime
   * strings are accepted at the type level and validated fail-closed at
   * runtime, so a JS caller threading an externally-resolved token gets the
   * structured rejection, not a type hole.
   */
  engine?: string;
  /**
   * Per-evaluation wall-clock budget (ms). Forwarded to the evaluator's
   * constructor with IDENTICAL semantics (positive integer,
   * <= MAX_TIMEOUT_MS); when omitted the evaluator's own DEFAULT_TIMEOUT_MS
   * governs, exactly as if the class had been constructed directly.
   */
  timeoutMs?: number;
  /**
   * Pure UDFs to register. Forwarded verbatim. The wasm engine accepts ONLY
   * the 3 native relay.* UDFs and rejects any other fail-closed
   * (RelayCelUnsupportedUdfError / RELAY-CEL-004).
   */
  udfs?: readonly PureUdf[];
}

/**
 * Resolve the engine token from the factory options (the TS counterpart of
 * engine.py _select_engine_name, minus the env read -- see the determinism
 * boundary in the module docstring).
 *
 * An absent or blank value resolves to the default (wasm). A non-blank value
 * is trimmed of surrounding whitespace (a common config-plumbing accident) and
 * matched case-sensitively against the (wasm-only) allowed token. A non-string
 * runtime value is a category error (TypeError); any unknown token -- including
 * the removed legacy spellings -- raises a clear Error naming the value and the
 * allowed set.
 */
function selectEngineToken(engine: string | undefined): typeof ENGINE_WASM {
  if (engine === undefined) {
    return DEFAULT_ENGINE;
  }
  if (typeof engine !== "string") {
    // A JS caller bypassing the types. Reject the category error loudly
    // rather than coercing (String(engine) could silently select an engine).
    throw new TypeError(
      `makeCelEvaluator: 'engine' must be a string when provided; ` +
        `got ${typeof engine}`,
    );
  }
  const value = engine.trim();
  if (value === "") {
    // A blank selection (e.g. an empty config field) is the standard "no
    // selection" signal; fall back to the default (wasm).
    return DEFAULT_ENGINE;
  }
  if (value === ENGINE_WASM) {
    return ENGINE_WASM;
  }
  const allowed = ALLOWED_ENGINE_TOKENS.map((token) =>
    JSON.stringify(token),
  ).join(", ");
  throw new Error(
    `makeCelEvaluator: engine ${JSON.stringify(engine)} is not a recognized ` +
      `CEL engine; allowed values are ${allowed} (unset or blank -> ` +
      `${JSON.stringify(DEFAULT_ENGINE)}). Engine names are case-sensitive. ` +
      "The legacy engine was removed at M6; the wasm engine is the only " +
      "CEL backend.",
  );
}

/**
 * Construct the CEL evaluator -- the package's canonical evaluator entry point
 * (VAL-CWC-P5FLIP-011 / VAL-CWC-P6REMOVE-005).
 *
 * `engine` unset / blank / `"wasm"` -> the wasm `WasmCelBackend` (the only
 * engine after M6 WS-I); any other token fails closed. `timeoutMs` and `udfs`
 * are forwarded to the evaluator's constructor with IDENTICAL semantics, so
 * the factory is a transparent substitute for constructing the class directly.
 *
 * The wasm backend's `evaluate()` is async (returns a Promise; the
 * worker_threads hard-kill timeout path).
 *
 * @throws TypeError `engine` is a non-string runtime value.
 * @throws Error `engine` holds an unrecognized token (incl. the removed legacy
 *   spellings), OR the forwarded `timeoutMs` is out of bounds (re-raised from
 *   the evaluator constructor).
 * @throws RelayCelUnsupportedUdfError a non-allowlist UDF was forwarded to the
 *   wasm engine (fail-closed at construction).
 */
export function makeCelEvaluator(
  options?: MakeCelEvaluatorOptions,
): WasmCelBackend {
  // `?? {}` (not a parameter default) so a JS caller passing an explicit
  // `null` gets the default-construction path rather than a bare property
  // read on null.
  const opts = options ?? {};
  // Validate the engine token fail-closed (unknown / legacy tokens throw here).
  selectEngineToken(opts.engine);

  // Forward only the options the caller actually supplied, so the evaluator's
  // own defaults govern the unspecified ones (identical to direct
  // construction). Mirrors engine.py's conditional kwargs build.
  const forwarded: { timeoutMs?: number; udfs?: readonly PureUdf[] } = {};
  if (opts.timeoutMs !== undefined) {
    forwarded.timeoutMs = opts.timeoutMs;
  }
  if (opts.udfs !== undefined) {
    forwarded.udfs = opts.udfs;
  }

  return new WasmCelBackend(forwarded);
}
